xds/internal/httpfilter/rbac/rbac.go

```diff
@@ -0,0 +1,181 @@
+		for _, principal := range policy.Principals {
+			if principal.GetHeader() != nil {
+				name := principal.GetHeader().GetName()
+				if name == ":scheme" || strings.HasPrefix(name, "grpc-") {
+					return nil, fmt.Errorf("rbac: principal header matcher for %v is :scheme or starts with grpc", name)
+				}
+			}
+		}
+		for _, permission := range policy.Permissions {
+			if permission.GetHeader() != nil {
+				name := permission.GetHeader().GetName()
+				if name == ":scheme" || strings.HasPrefix(name, "grpc-") {
+					return nil, fmt.Errorf("rbac: permission header matcher for %v is :scheme or starts with grpc", name)
+				}
+			}
+		}
+	}
+
+	return config{config: msg}, nil
+}
+
+func (builder) ParseFilterConfig(cfg proto.Message) (httpfilter.FilterConfig, error) {
+	return parseConfig(cfg)
+}
+
+func (builder) ParseFilterConfigOverride(override proto.Message) (httpfilter.FilterConfig, error) {
+	any, ok := override.(*anypb.Any)
+	if !ok {
+		return nil, fmt.Errorf("rbac: error parsing override config %v: unknown type %T", override, override)
+	}
+	msg := new(rpb.RBACPerRoute)
+	if err := ptypes.UnmarshalAny(any, msg); err != nil {
+		return nil, fmt.Errorf("rbac: error parsing override config %v: %v", override, err)
+	}
+	any, err := ptypes.MarshalAny(msg.Rbac)
+	if err != nil {
+		// This shouldn't happen.
+		return nil, fmt.Errorf("rbac: error marshaling override config %v", err)
+	}
+	return parseConfig(any)
+}
+
+func (builder) IsTerminal() bool {
+	return false
+}
+
+var _ httpfilter.ServerInterceptorBuilder = builder{}
+
+// BuildServerInterceptor is an optional interface builder implements in order
+// to signify it works server side.
+func (builder) BuildServerInterceptor(cfg httpfilter.FilterConfig, override httpfilter.FilterConfig) (resolver.ServerInterceptor, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("rbac: nil config provided")
+	}
+
+	c, ok := cfg.(config)
+	if !ok {
+		return nil, fmt.Errorf("rbac: incorrect config type provided (%T): %v", cfg, cfg)
+	}
+
+	if override != nil {
+		// override completely replaces the listener configuration; but we
+		// still validate the listener config type.
+		c, ok = override.(config)
+		if !ok {
+			return nil, fmt.Errorf("rbac: incorrect override config type provided (%T): %v", override, override)
+		}
+	}
+
+	icfg := c.config
+	// "If absent, no enforcing RBAC policy will be applied" - RBAC
+	// Documentation for Rules field.
+	if icfg.Rules == nil {
+		return &interceptor{}, nil
+	}
+
+	ce, err := rbac.NewChainEngine([]*v3rbacpb.RBAC{icfg.Rules})
+	if err != nil {
+		return nil, fmt.Errorf("error constructing matching engine: %v", err)
+	}
+	return &interceptor{chainEngine: ce}, nil
+}
+
+type interceptor struct {
+	chainEngine *rbac.ChainEngine
+}
+
+func (i *interceptor) AllowRPC(ctx context.Context) error {
+	// "If absent, no enforcing RBAC policy will be applied" - RBAC
+	// Documentation.
+	if i.chainEngine == nil {
```

Comment:
You won't need this check the the Build above returns `nil, nil`

Reply:
Deleted.